frame/pallet-xvm/src/wasm.rs

```diff
@@ -55,39 +56,42 @@ where
             error: XvmError::EncodingFailure,
             consumed_weight: PLACEHOLDER_WEIGHT,
         })?;
-        let res = pallet_contracts::Pallet::<T>::call(
-            frame_support::dispatch::RawOrigin::Signed(from).into(),
+
+        let dest = T::Lookup::lookup(dest).map_err(|error| XvmCallError {
+            error: XvmError::ExecutionError(Into::<&str>::into(error).into()),
+            consumed_weight: PLACEHOLDER_WEIGHT,
+        })?;
+        let call_result = pallet_contracts::Pallet::<T>::bare_call(
+            from, // no need to check origin, we consider it signed here
             dest,
             Default::default(),
             gas_limit.into(),
             None,
             input,
-        )
-        .map_err(|e| {
-            let consumed_weight = if let Some(weight) = e.post_info.actual_weight {
-                weight.ref_time()
-            } else {
-                gas_limit.ref_time()
-            };
-            XvmCallError {
-                error: XvmError::ExecutionError(Into::<&str>::into(e.error).into()),
-                consumed_weight,
-            }
-        })?;
+            false,
+            pallet_contracts::Determinism::Deterministic,
+        );
 
         log::trace!(
             target: "xvm::WASM::xvm_call",
-            "WASM XVM call result: {:?}", res
+            "WASM XVM call result: {:?}", call_result
         );
 
-        let consumed_weight = if let Some(weight) = res.actual_weight {
-            weight.ref_time()
-        } else {
-            gas_limit.ref_time()
-        };
-        Ok(XvmCallOk {
-            output: Default::default(), // TODO: Fill in with output from the call
-            consumed_weight,
-        })
+        let consumed_weight = call_result
+            .gas_consumed
+            .saturating_add(T::WeightInfo::call())
```

Comment:
Originally the code, instead of `bare_call`, contained a call to [the pallet's `call` method](https://github.com/paritytech/substrate/blob/master/frame/contracts/src/lib.rs#L635) that internally does 

```rust
			output.gas_meter.into_dispatch_result(output.result, T::WeightInfo::call())
```

which, in turn does the following:
```rust
	/// Turn this GasMeter into a DispatchResult that contains the actually used gas.
	pub fn into_dispatch_result<R, E>(
		self,
		result: Result<R, E>,
		base_weight: Weight,
	) -> DispatchResultWithPostInfo
	where
		E: Into<ExecError>,
	{
		let post_info = PostDispatchInfo {
			actual_weight: Some(self.gas_consumed().saturating_add(base_weight)),
			pays_fee: Default::default(),
		};

		result
			.map(|_| post_info)
			.map_err(|e| DispatchErrorWithPostInfo { post_info, error: e.into().error })
	}
```

It could be that I misinterpreted the logic here.

Reply:
@akru what's your thoughts on this?

Reply:
You should go through the code, and understand why that extra weight is added there, and why the _dry-run_ doesn't use it.

Reply:
This is not dry run. In pallet contracts, both `call` and `bare_call` use exactly the same `Self::internal_call`, but the former converts the result to dispatch result.

```rust
		#[pallet::call_index(6)]
		#[pallet::weight(T::WeightInfo::call().saturating_add(*gas_limit))]
		pub fn call(
			origin: OriginFor<T>,
			dest: AccountIdLookupOf<T>,
			#[pallet::compact] value: BalanceOf<T>,
			gas_limit: Weight,
			storage_deposit_limit: Option<<BalanceOf<T> as codec::HasCompact>::Type>,
			data: Vec<u8>,
		) -> DispatchResultWithPostInfo {
			let gas_limit: Weight = gas_limit.into();
			let origin = ensure_signed(origin)?;
			let dest = T::Lookup::lookup(dest)?;
			let mut output = Self::internal_call(
				origin,
				dest,
				value,
				gas_limit,
				storage_deposit_limit.map(Into::into),
				data,
				None,
				Determinism::Deterministic,
			);
			if let Ok(retval) = &output.result {
				if retval.did_revert() {
					output.result = Err(<Error<T>>::ContractReverted.into());
				}
			}
			output.gas_meter.into_dispatch_result(output.result, T::WeightInfo::call())
		}
```

```rust
	pub fn bare_call(
		origin: T::AccountId,
		dest: T::AccountId,
		value: BalanceOf<T>,
		gas_limit: Weight,
		storage_deposit_limit: Option<BalanceOf<T>>,
		data: Vec<u8>,
		debug: bool,
		determinism: Determinism,
	) -> ContractExecResult<BalanceOf<T>> {
		let mut debug_message = if debug { Some(DebugBufferVec::<T>::default()) } else { None };
		let output = Self::internal_call(
			origin,
			dest,
			value,
			gas_limit,
			storage_deposit_limit,
			data,
			debug_message.as_mut(),
			determinism,
		);
		ContractExecResult {
			result: output.result.map_err(|r| r.error),
			gas_consumed: output.gas_meter.gas_consumed(),
			gas_required: output.gas_meter.gas_required(),
			storage_deposit: output.storage_deposit,
			debug_message: debug_message.unwrap_or_default().to_vec(),
		}
	}
```

As I mentioned before, I did that addition to keep the behavior exactly the same as it was before the change.

Reply:
> This is not dry run. In pallet contracts, both call and bare_call use exactly the same Self::internal_call, but the former converts the result to dispatch result.

Please check how the `bare_call ` is used in `Astar` repo (runtime code).

> As I mentioned before, I did that addition to keep the behavior exactly the same as it was before the change.

And as I mentioned before so many times, we should understand WHY we're doing what we are doing.
Copy/pasting it from somewhere and saying that's the reason is not good enough.

This is the code from runtime code (for the dry-run):
```Rust
        fn call(
            origin: AccountId,
            dest: AccountId,
            value: Balance,
            gas_limit: Option<Weight>,
            storage_deposit_limit: Option<Balance>,
            input_data: Vec<u8>,
        ) -> pallet_contracts_primitives::ContractExecResult<Balance> {
            let gas_limit = gas_limit.unwrap_or(RuntimeBlockWeights::get().max_block);
            Contracts::bare_call(origin, dest, value, gas_limit, storage_deposit_limit, input_data, true, pallet_contracts::Determinism::Deterministic)
        }
```

Why isn't the extra weight added here as well?